[7.7.x] RHPAM-1393 & RHPAM-1406: Stunner & bendpoints issues

romartin opened the pull request:
Hey @hasys  @tiagodolphine 

Here are the fixes for two blockers issues:
* [RHPAM-1393](https://issues.jboss.org/browse/RHPAM-1393) - I've just cherry-picked it from master
* [RHPAM-1406](https://issues.jboss.org/browse/RHPAM-1406) - It fixes the issues related to connectors' bendpoints when doing multiple selection as well. So at this point, you can select any single or multiple bendpoints, or even the whole diagram as the ticket describes, and it updates the locations for everything

@tiagodolphine I changed the implementation that was done here... lemme know what you think please!!
* it was relying on add "custom" event handlers for the bendpoints for the lienzo connectors, and firing CDI events, which were consumed by stunner components in order to fire commands. 
* I refactored it because:
** It was not consistent with the other lienzo "hook" (like containment, connection acceptors)
** By firing events, instead of using the "hooks", I think it can result on un-ordered command executions or compositing
** This way created the `IControlPointsAcceptor` and related stuff
* Also updated the `UpdateControlPointPosition` commands - they're not adding/removing now, they're just updating the current points
Also you know details for this stuff, as you fixed some related issues before

It depends on:
https://github.com/kiegroup/lienzo-core/pull/55
https://github.com/kiegroup/lienzo-tests/pull/37

PS: I'm still missing to add more tests, please be patient on this :) But I need you to please start doing the review and testing of this ASAP. I'm leaving this week and this still needs to be integrated on master.

Thanks!

romartin commented:
Jenkins execute full downstream build

mareknovotny commented:
Jenkins execute full downstream build

hasys commented:
Jenkins retest please

tiagodolphine commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/command/DeleteCanvasControlPointCommand.java:
```diff
@@ -35,8 +35,8 @@ public DeleteCanvasControlPointCommand(final Edge candidate, final ControlPoint.
     public CommandResult<CanvasViolation> execute(AbstractCanvasHandler context) {
-        ShapeUtils.hideControlPoints(candidate, context);
         ShapeUtils.removeControlPoints(candidate, context, controlPoints);
+        ShapeUtils.hideControlPoints(candidate, context);
```
hide is after remove, doesn't it inverted?

tiagodolphine commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/main/java/org/kie/workbench/common/stunner/core/graph/command/impl/AddControlPointCommand.java:
```diff
@@ -35,18 +35,14 @@
     @Override
     protected CommandResult<RuleViolation> check(GraphCommandExecutionContext context) {
-        if (edge.getContent() instanceof HasControlPoints && validateControlPoints()) {
+        if (validateControlPoints()) {
```
minor, but using ternary here could be more clear.

tiagodolphine commented:
@romartin I found these issues I when moving sequence flows with CPs.
https://issues.jboss.org/secure/attachment/12438471/issue-moving-cp-1.webm
https://issues.jboss.org/secure/attachment/12438472/issue-moving-cp2.webm

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/menu/dev/impl/LogSelectedEdgeDevCommand.java:
```diff
@@ -0,0 +1,113 @@
+import org.kie.workbench.common.stunner.core.util.StunnerLogger;
+
+@Dependent
+public class LogSelectedEdgeDevCommand extends AbstractSelectionDevCommand {
```
All the XXXDevCommand's are actually only present in the showcase-standalone, this is because it uses a custom screen class that shows those items on the menu, On the other hand, the webapps based on KIE Editors, like the production ones, are not displaying them, so this actually "works".
The only downside here is that the code for those classes, as it's present in the widgets package, is being anyway exported, but we can improve this later on... we're talking about just a few code.. :)

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/shape/view/wires/WiresConnectorView.java:
```diff
@@ -103,55 +100,59 @@ public String getUUID() {
+    public T updateControlPoint(final ControlPoint controlPoint) {
+        if (validateControlPointShape()) {
+            final Point2D location = controlPoint.getLocation();
+            moveControlPoint(controlPoint.getIndex() + 1,
```
Yeah sure np, lemme explain why I did this.: 
* WiresConnectorView is a (implements) Stunner ShapeView - it acts as a mediator between both  Stunner's view  and Lienzo instances
* Also the CP in stunner are declared a bit different rather than on lienzo side
** On Lienzo the control handles at indexes 0 and size-1 are the ones used for magnets/connections
** On Stunner the connector has two (optional) connections and the CPs, as 2 different things
* This implies that a connector's CP for the index x links to the the control handle with index x+1 on the lienzo side
* Finally due to WiresConnectorView implements HasManageableControlPoints, which rely on the Stunner model, and it calls the related handling methods for CPs in the WiresConnector, I applied the logic about doing this index mapping here,...

As usually, just giving my opinion, but all comments and other ideas are welcome!! :) Thanks for the reviews anyway, are being really useful! :+1:

romartin commented:
Hey @hasys @tiagodolphine 

Thanks for the great review! :+1: 

I've added new commits that apply your comments from this PR and also fixes the bugs you both found related to bendpoints and parent shapes, sorry for this, was my mistake, it was missing to fire the acceptor controls for child shapes.

@hasys about the issues you mention above - I've fixed `1` and `2`, but `3` and `4` are not really related to this, are related to copy/paste stuff, and are not even introduced on this PR, they're present on current 7.0.2.CR1 release. Same for the "unrelated" issue you found as well :(  
So please, are you agree on opening new tickets for those issues? 

@hasys @tiagodolphine please give it a look... for now I understand I'm only missing to improve the test coverage, so working on it! If any issues lemme know and will fix.

Thanks!

romartin commented:
Jenkins execute full downstream build

hasys commented:
Hi @romartin,

great work, thank you Roger!

Just some notes:

* yes, sure I will report 3,4 and unrelated issues from previous review as separate Jiras.
* there are some failing Stunner tests, looks like caused by your changes (because I guess there are no other changes arrived to 7.7.x)
* I found two new issues (see below)
* there are still some missing test coverage (will provide more info where, but some of my previous comments on lienzo-core about missing coverage are still valid)
* I put a comment to lienzo-core about dead code

Issues:

**Not possible to move direction points in group**
- Select Sequence flow with direction points as group selection
- Move sequence flow

_Actual Result_
While you dragging all direction points are moving, when you release a button all points will be returned to initial position.
_Expected Result_
When you release mouse button all direction points will preserve it’s positioin

**Direction points out of container are moving when you move container**
- Create [Start→Task→End] process in a way that Start will be on process level and the rest of the process will be inside of swimlane. Sequence flow connecting Start with Task should have some direction points outside of swimlane.
- Move Swimlane

_Actual Result_
Direction points will move along with swimlane
_Expected Result_
Direction points will be “attached” to top level process.

romartin commented:
Closing as finally it only goes to `master` branch. So created https://github.com/kiegroup/kie-wb-common/pull/2004
